File: app/src/main/java/io/stamethyst/backend/LaunchPreparationService.java
package io.stamethyst.backend;

import android.content.Context;

import androidx.annotation.Nullable;

import java.io.IOException;

public final class LaunchPreparationService {
    private LaunchPreparationService() {
    }

    public static void prepare(Context context, String launchMode) throws IOException {
        prepare(context, launchMode, null);
    }

    public static void prepare(
            Context context,
            String launchMode,
            @Nullable StartupProgressCallback progressCallback
    ) throws IOException {
        reportProgress(progressCallback, 3, "Installing launcher components...");
        ComponentInstaller.ensureInstalled(context, mapProgressRange(progressCallback, 5, 35));

        reportProgress(progressCallback, 36, "Preparing Java runtime...");
        RuntimePackInstaller.ensureInstalled(context, mapProgressRange(progressCallback, 36, 76));

        reportProgress(progressCallback, 78, "Ensuring runtime directories...");
        RuntimePaths.ensureBaseDirs(context);

        reportProgress(progressCallback, 86, "Validating desktop-1.0.jar...");
        StsJarValidator.validate(RuntimePaths.importedStsJar(context));

        if (StsLaunchSpec.LAUNCH_MODE_MTS_BASEMOD.equals(launchMode)) {
            reportProgress(progressCallback, 90, "Validating required mod jars...");
            ModJarSupport.validateMtsJar(RuntimePaths.importedMtsJar(context));
            ModJarSupport.validateBaseModJar(RuntimePaths.importedBaseModJar(context));
            ModJarSupport.validateStsLibJar(RuntimePaths.importedStsLibJar(context));

            reportProgress(progressCallback, 95, "Preparing MTS classpath...");
            ModJarSupport.prepareMtsClasspath(context);
            ModManager.resolveLaunchModIds(context);
        }

        reportProgress(progressCallback, 100, "Launch preparation complete");
    }

    @Nullable
    private static StartupProgressCallback mapProgressRange(
            @Nullable StartupProgressCallback callback,
            int startPercent,
            int endPercent
    ) {
        if (callback == null) {
            return null;
        }
        int safeStart = clampPercent(startPercent);
        int safeEnd = clampPercent(endPercent);
        return (percent, message) -> callback.onProgress(
                mapRangeProgress(percent, safeStart, safeEnd),
                message
        );
    }

    private static int mapRangeProgress(int percent, int startPercent, int endPercent) {
        int bounded = clampPercent(percent);
        float ratio = bounded / 100f;
        return startPercent + Math.round((endPercent - startPercent) * ratio);
    }

    private static int clampPercent(int value) {
        return Math.max(0, Math.min(100, value));
    }

    private static void reportProgress(
            @Nullable StartupProgressCallback callback,
            int percent,
            String message
    ) {
        if (callback != null) {
            callback.onProgress(clampPercent(percent), message);
        }
    }
}

File: app/src/main/java/io/stamethyst/backend/launch/LaunchPreparationService.kt
package io.stamethyst.backend.launch

import android.content.Context
import io.stamethyst.R
import io.stamethyst.backend.diag.MemoryDiagnosticsLogger
import io.stamethyst.backend.launch.ComponentInstaller
import io.stamethyst.config.RuntimePaths
import io.stamethyst.backend.mods.ModJarSupport
import io.stamethyst.backend.mods.ModManager
import io.stamethyst.backend.mods.OptionalModStorageCoordinator
import io.stamethyst.backend.mods.StsJarValidator
import io.stamethyst.backend.resources.ExternalResourcePackService
import io.stamethyst.backend.runtime.RuntimePackInstaller
import java.io.IOException
import kotlin.math.roundToInt

object LaunchPreparationService {
    @Throws(IOException::class)
    private fun throwIfInterrupted() {
        if (Thread.currentThread().isInterrupted) {
            throw IOException("Launch preparation cancelled")
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun prepare(context: Context, launchMode: String) {
        prepare(context, launchMode, null)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun prepare(
        context: Context,
        launchMode: String,
        progressCallback: StartupProgressCallback?
    ): ModManager.LaunchModSnapshot? {
        MemoryDiagnosticsLogger.logEvent(
            context,
            "launch_preparation_started",
            mapOf("launchMode" to launchMode)
        )
        throwIfInterrupted()
        reportProgress(
            progressCallback,
            2,
            context.progressText(R.string.startup_progress_checking_external_resources)
        )
        ExternalResourcePackService.ensureAvailable(context, mapProgressRange(progressCallback, 2, 20))

        throwIfInterrupted()
        reportProgress(
            progressCallback,
            21,
            context.progressText(R.string.startup_progress_installing_launcher_components)
        )
        ComponentInstaller.ensureInstalled(context, mapProgressRange(progressCallback, 21, 43))
        ExternalResourcePackService.installNativeLibraries(context)

        throwIfInterrupted()
        reportProgress(
            progressCallback,
            44,
            context.progressText(R.string.startup_progress_preparing_java_runtime)
        )
        RuntimePackInstaller.ensureInstalled(context, mapProgressRange(progressCallback, 44, 76))

        throwIfInterrupted()
        reportProgress(
            progressCallback,
            78,
            context.progressText(R.string.startup_progress_ensuring_runtime_directories)
        )
        RuntimePaths.ensureBaseDirs(context)

        throwIfInterrupted()
        reportProgress(
            progressCallback,
            86,
            context.progressText(R.string.startup_progress_validating_desktop_jar)
        )
        StsJarValidator.validate(RuntimePaths.importedStsJar(context))

        if (StsLaunchSpec.isMtsLaunchMode(launchMode)) {
            throwIfInterrupted()
            reportProgress(
                progressCallback,
                90,
                context.progressText(R.string.startup_progress_validating_required_mod_jars)
            )
            ModJarSupport.validateMtsJar(RuntimePaths.importedMtsJar(context))
            ModJarSupport.validateBaseModJar(RuntimePaths.importedBaseModJar(context))
            ModJarSupport.validateStsLibJar(RuntimePaths.importedStsLibJar(context))

            throwIfInterrupted()
            reportProgress(
                progressCallback,
                96,
                context.progressText(R.string.startup_progress_resolving_enabled_mod_launch_list)
            )
            val launchSnapshot = ModManager.buildLaunchModSnapshot(context)
            OptionalModStorageCoordinator.prepareMtsModFileList(context, launchSnapshot)
            MemoryDiagnosticsLogger.logModSnapshot(
                context = context,
                event = "launch_preparation_resolved_launch_mods",
                launchMode = launchMode,
                enabledLibraryFiles = launchSnapshot.enabledLibraryFiles,
                runtimeModFiles = launchSnapshot.launchModFiles,
                launchModIds = launchSnapshot.launchModIds
            )
            finishPreparation(context, launchMode, progressCallback)
            return launchSnapshot
        }

        finishPreparation(context, launchMode, progressCallback)
        return null
    }

    private fun finishPreparation(
        context: Context,
        launchMode: String,
        progressCallback: StartupProgressCallback?
    ) {
        throwIfInterrupted()
        reportProgress(
            progressCallback,
            100,
            context.progressText(R.string.startup_progress_launch_preparation_complete)
        )
        MemoryDiagnosticsLogger.logEvent(
            context,
            "launch_preparation_completed",
            mapOf("launchMode" to launchMode)
        )
    }

    private fun mapProgressRange(
        callback: StartupProgressCallback?,
        startPercent: Int,
        endPercent: Int
    ): StartupProgressCallback? {
        if (callback == null) {
            return null
        }
        val safeStart = clampPercent(startPercent)
        val safeEnd = clampPercent(endPercent)
        return StartupProgressCallback { percent, message ->
            callback.onProgress(mapRangeProgress(percent, safeStart, safeEnd), message)
        }
    }

    private fun mapRangeProgress(percent: Int, startPercent: Int, endPercent: Int): Int {
        val bounded = clampPercent(percent)
        val ratio = bounded / 100f
        return startPercent + ((endPercent - startPercent) * ratio).roundToInt()
    }

    private fun clampPercent(value: Int): Int {
        return value.coerceIn(0, 100)
    }

    private fun reportProgress(callback: StartupProgressCallback?, percent: Int, message: String) {
        callback?.onProgress(clampPercent(percent), message)
    }
}
